demos/mqtt/mqtt_demo_basic_tls/mqtt_demo_basic_tls.c

```diff
@@ -0,0 +1,470 @@
+
+/**
+ * @brief Set up a TLS connection over an existing TCP connection.
+ *
+ * @param[in] tcpSocket Existing TCP connection.
+ *
+ * @return An SSL connection context; NULL on failure.
+ */
+static SSL * tlsSetup( int tcpSocket )
+{
+    int sslStatus = 0;
+    FILE * pRootCaFile = NULL;
+    X509 * pRootCa = NULL;
+    SSL * pSslContext = NULL;
+
+    /* Read the server root CA certificate. */
+    SSL_CTX * pSslSetup = SSL_CTX_new( TLS_client_method() );
+
+    if( pSslSetup != NULL )
+    {
+        /* Set auto retry mode for the blocking calls to SSL_read and SSL_write.
+         * The mask returned by SSL_CTX_set_mode does not need to be checked. */
+        ( void ) SSL_CTX_set_mode( pSslSetup, SSL_MODE_AUTO_RETRY );
+
+        /* OpenSSL does not provide a single function for reading and loading certificates
+         * from files into stores, so the file API must be called. */
+        pRootCaFile = fopen( SERVER_CERT, "r" );
+
+        if( pRootCaFile != NULL )
+        {
+            pRootCa = PEM_read_X509( pRootCaFile, NULL, NULL, NULL );
+        }
+
+        if( pRootCa != NULL )
+        {
+            sslStatus = X509_STORE_add_cert( SSL_CTX_get_cert_store( pSslSetup ),
+                                               pRootCa );
+        }
+    }
+
+    /* Set up the TLS connection. */
+    if( sslStatus == 1 )
+    {
+        /* Create a new SSL context. */
+        pSslContext = SSL_new( pSslSetup );
+
+        if( pSslContext != NULL )
+        {
+            /* Enable SSL peer verification. */
+            SSL_set_verify( pSslContext, SSL_VERIFY_PEER, NULL );
+
+            sslStatus = SSL_set_fd( pSslContext, tcpSocket );
+        }
+        else
+        {
+            sslStatus = 0;
+        }
+
+        /* Perform the TLS handshake. */
+        if( sslStatus == 1 )
+        {
+            sslStatus = SSL_connect( pSslContext );
+        }
+
+        if( sslStatus == 1 )
+        {
+            if( SSL_get_verify_result( pSslContext ) != X509_V_OK )
+            {
+                sslStatus = 0;
+            }
+        }
+
+        /* Clean up on error. */
+        if( sslStatus == 0 )
+        {
+            SSL_free( pSslContext );
+            pSslContext = NULL;
+        }
+    }
+
+    if( pRootCaFile != NULL )
+    {
+        ( void ) fclose( pRootCaFile );
+    }
+
+    if( pRootCa != NULL )
+    {
+        X509_free( pRootCa );
+    }
+
+    if( pSslSetup != NULL )
+    {
+        SSL_CTX_free( pSslSetup );
+    }
+
+    return pSslContext;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief The transport send function provided to the MQTT context.
+ *
+ * @param[in] pSslContext SSL context.
+ * @param[in] pMessage Data to send.
+ * @param[in] bytesToSend Length of data to send.
+ *
+ * @return Number of bytes sent; negative value on error.
+ */
+static int32_t transportSend( SSL * pSslContext, const void * pMessage, size_t bytesToSend )
```

Comment:
I see the `typedef` but to ensure that the signature matches (while reading) as mentioned in the MQTT headers, do you think it is better to use `MQTTNetworkContext_t` instead of `SSL *`?

Reply:
Fixed